File: backend/app/routers/agent.py
# -*- coding: utf-8 -*-
"""Agent 基础接口：工具发现、只读调用和审计查看。"""
from __future__ import annotations

import json

from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from ..agent.agent_service import invoke_tool, list_audits, list_tools
from ..agent.model_config import load_local_config, public_config, save_local_config
from ..agent.model_client import ModelError, ModelConfig
from ..agent.runner import AgentRunner
from ..agent.session_store import SessionStore
from ..agent.tool_registry import ToolError
from ..wechat.service import wechat_service

router = APIRouter(prefix='/api/agent')


class ToolExecuteBody(BaseModel):
    arguments: dict = Field(default_factory=dict)
    channel: str = 'local'
    actor_id: str = ''


class ChatBody(BaseModel):
    session_id: str = Field(min_length=1, max_length=120)
    message: str = Field(min_length=1, max_length=4000)
    channel: str = 'local'
    actor_id: str = ''


class ModelConfigBody(BaseModel):
    api_key: str | None = None
    base_url: str = Field(min_length=1, max_length=500)
    model: str = Field(min_length=1, max_length=120)
    thinking: str = 'disabled'
    clear_api_key: bool = False


@router.get('/status')
def agent_status():
    config = ModelConfig.from_env()
    wechat = wechat_service.status()
    return {
        'enabled': True,
        'model': config.model or 'not_configured',
        'model_configured': config.configured,
        'wechat': 'iLink',
        'wechat_configured': wechat['configured'],
        'wechat_running': wechat['running'],
        'tool_count': len(list_tools()),
        'message': 'Agent 工具、模型客户端和微信接入接口已就绪，是否启用取决于本地配置。',
    }


@router.get('/tools')
def agent_tools():
    return {'tools': list_tools()}


@router.get('/config')
def agent_config():
    local = load_local_config()
    effective = ModelConfig.from_env()
    data = public_config({
        'api_key': effective.api_key,
        'base_url': effective.base_url,
        'model': effective.model,
        'thinking': effective.thinking,
    })
    data['storage'] = 'local'
    data['local_override_active'] = bool(local)
    return data


@router.put('/config')
def update_agent_config(body: ModelConfigBody):
    if body.thinking not in {'disabled', 'enabled'}:
        raise HTTPException(400, 'thinking 只能是 disabled 或 enabled')
    values = load_local_config()
    if body.clear_api_key:
        values.pop('api_key', None)
    elif body.api_key:
        values['api_key'] = body.api_key.strip()
    values.update({
        'base_url': body.base_url.rstrip('/'),
        'model': body.model.strip(),
        'thinking': body.thinking,
    })
    save_local_config(values)
    return {'ok': True, **public_config(values), 'storage': 'local'}


@router.post('/tools/{tool_name}')
def execute_agent_tool(tool_name: str, body: ToolExecuteBody):
    try:
        result = invoke_tool(
            tool_name,
            body.arguments,
            channel=body.channel,
            actor_id=body.actor_id,
        )
        return {'ok': True, 'tool': tool_name, 'result': result}
    except ToolError as exc:
        raise HTTPException(400, str(exc)) from exc


@router.post('/chat')
async def agent_chat(body: ChatBody):
    try:
        answer = await AgentRunner(session_store=SessionStore()).chat(
            body.session_id,
            body.message,
            channel=body.channel,
            actor_id=body.actor_id,
        )
        return {'ok': True, 'session_id': body.session_id, 'answer': answer}
    except ModelError as exc:
        raise HTTPException(503, str(exc)) from exc


@router.post('/chat/stream')
async def agent_chat_stream(body: ChatBody):
    async def events():
        try:
            runner = AgentRunner(session_store=SessionStore())
            async for event in runner.chat_stream(
                body.session_id,
                body.message,
                channel=body.channel,
                actor_id=body.actor_id,
            ):
                yield f"data: {json.dumps(event, ensure_ascii=False)}\n\n"
        except ModelError as exc:
            yield f"data: {json.dumps({'type': 'error', 'message': str(exc)}, ensure_ascii=False)}\n\n"
        except Exception:
            yield f"data: {json.dumps({'type': 'error', 'message': 'Agent 流式响应失败，请稍后重试。'}, ensure_ascii=False)}\n\n"
        yield 'data: {"type":"done"}\n\n'

    return StreamingResponse(
        events(),
        media_type='text/event-stream',
        headers={'Cache-Control': 'no-cache', 'X-Accel-Buffering': 'no'},
    )


@router.delete('/sessions/{session_id}')
def clear_agent_session(session_id: str):
    SessionStore().clear(session_id)
    return {'ok': True}


@router.get('/audit')
def agent_audit(limit: int = Query(50, ge=1, le=200)):
    return {'audits': list_audits(limit)}

File: server/src/http/routes/agent.ts
/* Agent 基础路由：状态、配置、工具、对话（含流式）、会话、审计与用量。
 * 全量会话管理与确认写入在 AGENT-02/03 扩展；本文件先提供可运行子集。
 */
import type { FastifyInstance, FastifyReply } from 'fastify';

import { AgentRunner } from '../../agent/runner.js';
import { pendingForSession, confirmAction, cancelAction, ActionError } from '../../agent/actions.js';
import { SessionStore } from '../../agent/sessionStore.js';
import { listTools, invokeTool, listAudits, usageStats } from '../../agent/agentService.js';
import { ToolError } from '../../agent/toolRegistry.js';
import { loadConfig, saveConfig, publicConfig, ModelConfigError } from '../../agent/modelConfig.js';
import { ModelError, ModelNotConfigured } from '../../agent/modelClient.js';

function mapAgentError(reply: FastifyReply, error: unknown): FastifyReply | undefined {
  if (error instanceof ModelNotConfigured || error instanceof ModelError || error instanceof ModelConfigError) {
    return reply.status(400).send({ detail: (error as Error).message });
  }
  if (error instanceof ToolError) {
    return reply.status(400).send({ detail: error.message, code: error.code });
  }
  return undefined;
}

function toolErrorDetail(error: unknown): { message: string; code?: string } | null {
  if (error instanceof ToolError) return { message: error.message, code: error.code };
  return null;
}

export function registerAgentRoutes(app: FastifyInstance): void {
  app.get('/api/agent/status', async () => ({
    enabled: true,
    configured: Boolean(loadConfig().api_key || process.env.MEIMEI_MODEL_API_KEY),
    model: loadConfig().model,
    model_configured: Boolean(loadConfig().api_key || process.env.MEIMEI_MODEL_API_KEY),
  }));

  app.get('/api/agent/config', async (request, reply) => {
    const host = request.ip;
    if (host !== '127.0.0.1' && host !== '::1' && host !== 'localhost' && host !== 'testclient') {
      return reply.status(403).send({ detail: '模型配置只能在工作台本机管理' });
    }
    return publicConfig(loadConfig() as unknown as Record<string, unknown>);
  });

  app.put('/api/agent/config', async (request, reply) => {
    const host = request.ip;
    if (host !== '127.0.0.1' && host !== '::1' && host !== 'localhost' && host !== 'testclient') {
      return reply.status(403).send({ detail: '模型配置只能在工作台本机管理' });
    }
    const body = request.body as Record<string, unknown>;
    try {
      return saveConfig({
        api_key: body.api_key === undefined ? null : String(body.api_key),
        base_url: String(body.base_url ?? ''),
        model: String(body.model ?? ''),
        thinking: String(body.thinking ?? 'disabled'),
        clear_api_key: body.clear_api_key === true,
      });
    } catch (error) {
      const mapped = mapAgentError(reply, error);
      if (mapped) return mapped;
      throw error;
    }
  });

  app.get('/api/agent/tools', async (request) => {
    const { channel = 'local' } = request.query as { channel?: string };
    return { tools: listTools(channel) };
  });

  app.post('/api/agent/tools/:toolName', async (request, reply) => {
    const { toolName } = request.params as { toolName: string };
    const body = request.body as { arguments?: Record<string, unknown>; channel?: string; actor_id?: string; session_id?: string };
    try {
      const result = invokeTool(toolName, body.arguments ?? {}, {
        channel: body.channel ?? 'local', actorId: body.actor_id ?? '',
        sessionId: body.session_id ?? '',
      });
      return result;
    } catch (error) {
      const mappedError = toolErrorDetail(error);
      if (mappedError) return reply.status(400).send(mappedError);
      throw error;
    }
  });

  app.post('/api/agent/chat', async (request, reply) => {
    const body = request.body as { session_id: string; message: string; channel?: string; actor_id?: string };
    try {
      const runner = new AgentRunner({ sessionStore: new SessionStore() });
      const answer = await runner.chat(body.session_id, body.message, {
        channel: body.channel ?? 'local', actorId: body.actor_id ?? '',
      });
      return { answer, session_id: body.session_id };
    } catch (error) {
      const mapped = mapAgentError(reply, error);
      if (mapped) return mapped;
      throw error;
    }
  });

  app.post('/api/agent/chat/stream', async (request, reply) => {
    const body = request.body as { session_id: string; message: string; channel?: string; actor_id?: string };
    reply.raw.writeHead(200, {
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      Connection: 'keep-alive',
    });
    const runner = new AgentRunner({ sessionStore: new SessionStore() });
    try {
      for await (const event of runner.chatStream(body.session_id, body.message, {
        channel: body.channel ?? 'local', actorId: body.actor_id ?? '',
      })) {
        reply.raw.write(`data: ${JSON.stringify(event)}\n\n`);
      }
    } catch (error) {
      const message = (error as Error).message;
      reply.raw.write(`data: ${JSON.stringify({ type: 'error', content: message })}\n\n`);
    } finally {
      reply.raw.end();
    }
    return reply;
  });

  app.get('/api/agent/sessions', async (request) => {
    const { prefix = '' } = request.query as { prefix?: string };
    return { sessions: new SessionStore().list(prefix) };
  });

  app.get('/api/agent/sessions/:sessionId', async (request, reply) => {
    const { sessionId } = request.params as { sessionId: string };
    const messages = new SessionStore().load(sessionId);
    if (!messages.length) return reply.status(404).send({ detail: '会话不存在' });
    return { messages };
  });

  app.put('/api/agent/sessions/:sessionId', async (request, reply) => {
    const { sessionId } = request.params as { sessionId: string };
    const body = request.body as { title?: string };
    try {
      const result = new SessionStore().rename(sessionId, String(body.title ?? ''));
      return result;
    } catch (error) {
      return reply.status(400).send({ detail: (error as Error).message });
    }
  });

  app.delete('/api/agent/sessions/:sessionId', async (request) => {
    const { sessionId } = request.params as { sessionId: string };
    new SessionStore().clear(sessionId);
    return { ok: true };
  });

  app.get('/api/agent/actions/pending', async (request) => {
    const { session_id, actor_id } = request.query as { session_id?: string; actor_id?: string };
    const pending = pendingForSession(session_id ?? '', actor_id ?? '');
    if (!pending) return { pending: null };
    return { pending };
  });

  app.post('/api/agent/actions/:actionId/confirm', async (request, reply) => {
    const { actionId } = request.params as { actionId: string };
    const body = request.body as { session_id?: string; actor_id?: string; confirmation_token?: string };
    try {
      const result = confirmAction(Number(actionId), {
        sessionId: body.session_id ?? '', actorId: body.actor_id ?? '',
        token: body.confirmation_token ?? '',
      });
      return { ok: true, ...result };
    } catch (error) {
      if (error instanceof ActionError) {
        return reply.status(400).send({ detail: error.message });
      }
      throw error;
    }
  });

  app.post('/api/agent/actions/:actionId/cancel', async (request, reply) => {
    const { actionId } = request.params as { actionId: string };
    const body = request.body as { session_id?: string; actor_id?: string };
    try {
      return cancelAction(Number(actionId), {
        sessionId: body.session_id ?? '', actorId: body.actor_id ?? '',
      });
    } catch (error) {
      if (error instanceof ActionError) {
        return reply.status(400).send({ detail: error.message });
      }
      throw error;
    }
  });

  app.get('/api/agent/audit', async (request) => {
    const { limit = '50' } = request.query as { limit?: string };
    return { items: listAudits(Number(limit)) };
  });

  app.get('/api/agent/usage', async () => usageStats());
}
